xcm/src/v3/junction.rs

```diff
@@ -0,0 +1,205 @@
+pub enum Junction {
+	/// An indexed parachain belonging to and operated by the context.
+	///
+	/// Generally used when the context is a Polkadot Relay-chain.
+	Parachain(#[codec(compact)] u32),
+	/// A 32-byte identifier for an account of a specific network that is respected as a sovereign endpoint within
+	/// the context.
+	///
+	/// Generally used when the context is a Substrate-based chain.
+	AccountId32 { network: Option<NetworkId>, id: [u8; 32] },
+	/// An 8-byte index for an account of a specific network that is respected as a sovereign endpoint within
+	/// the context.
+	///
+	/// May be used when the context is a Frame-based chain and includes e.g. an indices pallet.
+	AccountIndex64 {
+		network: Option<NetworkId>,
+		#[codec(compact)]
+		index: u64,
+	},
+	/// A 20-byte identifier for an account of a specific network that is respected as a sovereign endpoint within
+	/// the context.
+	///
+	/// May be used when the context is an Ethereum or Bitcoin chain or smart-contract.
+	AccountKey20 { network: Option<NetworkId>, key: [u8; 20] },
+	/// An instanced, indexed pallet that forms a constituent part of the context.
+	///
+	/// Generally used when the context is a Frame-based chain.
+	PalletInstance(u8),
+	/// A non-descript index within the context location.
+	///
+	/// Usage will vary widely owing to its generality.
+	///
+	/// NOTE: Try to avoid using this and instead use a more specific item.
+	GeneralIndex(#[codec(compact)] u128),
+	/// A nondescript datum acting as a key within the context location.
+	///
+	/// Usage will vary widely owing to its generality.
+	///
+	/// NOTE: Try to avoid using this and instead use a more specific item.
+	GeneralKey(Vec<u8>),
+	/// The unambiguous child.
+	///
+	/// Not currently used except as a fallback when deriving context.
+	OnlyChild,
+	/// A pluralistic body existing within consensus.
+	///
+	/// Typical to be used to represent a governance origin of a chain, but could in principle be used to represent
+	/// things such as multisigs also.
+	Plurality { id: BodyId, part: BodyPart },
+	/// A global network capable of externalizing its own consensus. This is not generally
+	/// meaningful outside of the universal level.
+	GlobalConsensus(NetworkId),
+}
+
+impl From<NetworkId> for Junction {
+	fn from(n: NetworkId) -> Self {
+		Self::GlobalConsensus(n)
+	}
+}
+
+impl From<[u8; 32]> for Junction {
+	fn from(id: [u8; 32]) -> Self {
+		Self::AccountId32 { network: None, id }
+	}
+}
+
+impl From<[u8; 20]> for Junction {
+	fn from(key: [u8; 20]) -> Self {
+		Self::AccountKey20 { network: None, key }
+	}
+}
+
+impl From<u64> for Junction {
+	fn from(index: u64) -> Self {
+		Self::AccountIndex64 { network: None, index }
+	}
+}
+
+impl From<u128> for Junction {
+	fn from(id: u128) -> Self {
+		Self::GeneralIndex(id)
+	}
+}
+
+impl From<Vec<u8>> for Junction {
+	fn from(id: Vec<u8>) -> Self {
+		Self::GeneralKey(id)
+	}
+}
+
+impl TryFrom<OldJunction> for Junction {
```

Comment:
Does this need to be a `TryFrom` when the function never returns an Err variant?

Reply:
Earlier versions of the function were not infallible, and over time as more changes creep in, these converters have a habit of becoming fallible. So it's easiest to keep it as a `TryFrom` for now.